### PR TITLE
crypto: Use a different library for ed25519/sr25519

### PR DESCRIPTION
**Pull request by Yawning**
At Oasis we have spend some time writing a new Ed25519/X25519/sr25519 implementation called curve25519-voi.  This PR switches the import from ed25519consensus/go-schnorrkel, which should lead to performance gains on most systems.

Summary of changes:
 * curve25519-voi is now used for Ed25519 operations, following the existing ZIP-215 semantics.
 * curve25519-voi's public key cache is enabled (hardcoded size of 4096 entries, should be tuned, see the code comment) to accelerate repeated Ed25519 verification with the same public key(s).
 * (BREAKING) curve25519-voi is now used for sr25519 operations.  This is a breaking change as the current sr25519 support does something decidedly non-standard when going from a MiniSecretKey to a SecretKey and or PublicKey (The expansion routine is called twice).  While I believe the new behavior (that expands once and only once) to be more "correct", this changes the semantics as implemented.
 * curve25519-voi is now used for merlin since the included STROBE implementation produces much less garbage on the heap.

Side issues fixed:
 * The version of go-schnorrkel that is currently imported by tendermint has a badly broken batch verification implementation.  Upstream has fixed the issue after I reported it, so the version should be bumped in the interim.

Open design questions/issues:
 * As noted, the public key cache size should be tuned.  It is currently backed by a trivial thread-safe LRU cache, which is not scan-resistant, but replacing it with something better is a matter of implementing an interface.
 * As far as I can tell, the only reason why serial verification on batch failure is necessary is to provide more detailed error messages (that are only used in some unit tests).  If you trust the batch verification to be consistent with serial verification then the fallback can be eliminated entirely (the BatchVerifier provided by the new library supports an option that omits the fallback if this is chosen as the way forward).
 * curve25519-voi's sr25519 support could use more optimization and more eyes on the code.  The algorithm unfortunately is woefully under-specified, and the implementation was done primarily because I got really sad when I actually looked at go-schnorrkel, and we do not use the algorithm at this time.

### CODE REVIEW

File: abci/types/pubkey.go
Comment (marbar3778) [inline abci/types/pubkey.go:42]: This shouldnt be necessary. The key gets passed as bytes.

Comment (Yawning) [reply abci/types/pubkey.go:42]: Hm, encoding.PubKeyToProto expects a crypto.PubKey, which in the case of sr25519, isn't a byte slice...

Comment (marbar3778) [reply abci/types/pubkey.go:42]: oh okay, it should be fine to keep it like this then

File: crypto/sr25519/batch.go
Comment (marbar3778) [inline crypto/sr25519/batch.go:11]: why remove?

File: crypto/sr25519/encoding.go
Comment (marbar3778) [inline crypto/sr25519/encoding.go:8]: ditto

Comment (marbar3778) [reply crypto/sr25519/encoding.go:8]: whoops looked at the next file

File: crypto/sr25519/batch.go
Comment (Yawning) [reply crypto/sr25519/batch.go:11]: See line 192 (It could be moved up, but not implementing the interface will cause a build failure anyway).

File: crypto/sr25519/encoding.go
Comment (Yawning) [reply crypto/sr25519/encoding.go:8]: Moved to pubkey.go/privkey.go since that's where the types are defined.

File: crypto/sr25519/batch.go
Comment (Yawning) [reply crypto/sr25519/batch.go:11]: Guess it got lost in the rewrite, I can re-add it.  Though the ctor returning the interface is sufficient to cause a build failure.

File: abci/types/pubkey.go
Comment (Yawning) [reply abci/types/pubkey.go:42]: I would be ok with switching sr25519.PubKey back to being `[]byte` since it would remove the need for the s11n related changes.  Ed25519 takes the hit of doing the decompression (but has a cache), each verification and performance is adequate.

The code is probably simpler to reason about that way...

Comment (Yawning) [reply abci/types/pubkey.go]: Changed it back, since it's simpler.  Eventually I'll just add a sr25519 public key cache.

File: crypto/sr25519/batch.go
Comment (Yawning) [reply crypto/sr25519/batch.go:11]: Re-added.

File: crypto/sr25519/privkey.go
Comment (Yawning) [reply crypto/sr25519/privkey.go]: Stray comment I left to remind myself to implement it, removed.

**Comment by ValarDragon** [comment]
Very excited for these to land! I feel like it would be nice to not have state machine tx pubkeys have risk of being cached / throwing out validator pubkeys. I don't think this should block merge, as current cosmos chains don't use Ed25519 for user signatures. One way this could be changed is by making a separate CachedVerify and Verify methods, or by making a cached pubkey struct, and have it managed at that layer, instead of managed via a hashmap on the back-end.

I'll try to review the cryptography during the week.

**Comment by marbar3778** [comment]
> I feel like it would be nice to not have state machine tx pubkeys have risk of being cached / throwing out validator pubkeys. I don't think this should block merge, as current cosmos chains don't use Ed25519 for user signatures

Cosmos-sdk has their own version of ed25519. We now look at the the Tendermint crypto package as solely tendermint related

**Comment by Yawning** [comment]
With the way I integrated support for the cache (lookup done at verification time), the overhead of a miss isn't terrible.  The extra computational overhead imposed by the caching layer is confined to synchronization (in the form of two tiny critical sections), book keeping, and allocating the cache entry, everything else is (mostly) work that has to happen to verify anyway.  It does also result in slightly more work for the GC since intermediaries that were stack allocated are now heap allocated.

In comparison to not using the cache at all, a single-verification will take a penalty of ~3.6 usec on my laptop, which while not ideal isn't what I consider to be the end of the world.

If this was to be more optimized, replacing the cache implementation with something tailored to the sort of access patterns that tendermint is expected to see should be simple (eg: have a cache that has a split backing store, with an LRU cache and a separate map, the latter which is manually managed populated on validator set changes).

**Comment by ValarDragon** [comment]
Oh thats not bad at all for cache miss times, seems good to me then

> Cosmos-sdk has their own version of ed25519. We now look at the the Tendermint crypto package as solely tendermint related

Ah thanks, didn't know this

File: crypto/ed25519/ed25519.go
Comment (marbar3778) [inline crypto/ed25519/ed25519.go:52]: We have an artificial limit of 10.000 validators (https://github.com/tendermint/tendermint/blob/HEAD/types/vote_set.go?q=maxvalset#L18). Should we open an issue with this? I don't think this will come up for a while because there are other limitations that would prevent 4096 validators.

**Review by marbar3778 — APPROVED** [review]
utACK. Would love another approval before approving

**Review by cmwaters — APPROVED** [review]
I have next to zero knowledge on crypto primitives so we might want a second pair of eyes on that but as to the verify functions - it looks all good to me.

**Comment by Yawning** [comment]
Just as a heads up, when I do the final rebase, I would like to bump the import to include https://github.com/oasisprotocol/curve25519-voi/pull/63.

The current STROBE implementation that is the same as the one pulled in by the existing merlin import causes most sr25519 operations to do horrific things to the heap, which I solved by writing a "from the specification" implementation of STROBE.  Details in the PR.

**Comment by Yawning** [comment]
Ok, I went ahead and rebased the branch, and bumped the import to pickup the STROBE changes.  As an added bonus I exposed the merlin implementation (forked from George Tankersley's) that I initially left as an internal package and switched the import there as well.

It is API incompatible because ExtractBytes takes a destination buffer, and labels are strings, but I view both changes as improvements since they both make the code nicer to read.

**Comment by Yawning** [comment]
Nudging this because my understanding is that this will move forward next week-ish.  Please let me know if I should rebase it to resolve the conflicts or anything, and as always please feel free to ask if you have questions.

**Comment by marbar3778** [comment]
@Yawning sorry for the delay. Could you resolve the conflicts and we can get to merging

**Comment by marbar3778** [comment]
oh whoops, also a changelog entry. Sorry, just noticed this. 

Can it mention the breaking changes as well?

cc @sunnya97 we are breaking sr25519 in 0.35. Do you need anything from us for straight edge.

**Comment by Yawning** [comment]
> oh whoops, also a changelog entry. Sorry, just noticed this.
> 
> Can it mention the breaking changes as well?

This should be done now.

> cc @sunnya97 we are breaking sr25519 in 0.35. Do you need anything from us for straight edge.

If they do not use `sr25519.GenPrivKeyFromSecret` things might be "ok", since I kept the byte-serialized `PrivKey` as a sr25519 MiniSecretKey.  I would still recommend existing keys that were generated with that routine to be regenerated because the extra expansion removes 3-bits of the scalar.

**Comment by marbar3778** [comment]
@Yawning sorry to bother again. Can you update to master again, then the bot should be able to merge this.

**Pull request merged**
